Upgrades TypeScript docs to 4.7 (#2380)
microsoft/TypeScript-Website: packages/documentation/copy/en/reference/ESM Support for Node.md

---
title: ECMAScript Modules in Node.js
layout: docs
permalink: /docs/handbook/esm-node.html
oneline: Using ECMAScript Modules in Node.js
---

For the last few years, Node.js has been working to support running ECMAScript modules (ESM).

> This has been a vesry difficult feature to support, since the foundation of the Node.js ecosystem is built on a different module system called CommonJS (CJS).

Reviewer: Noticed a typo here from the Twitter thread! 😅  `vesry` -> `very`

Author: Thanks!

Seamless tooling between the two runtime environments brings large challenges, with many new features to juggle;

> however, support for ESM in Node.js is now implemented in Node.js 12 and later, and the dust has begun to settle.

Reviewer 2: This sentence is still true, but it’s kind of weird to mention Node 12 and then follow up with the setting `node16`... maybe just leave it out

```suggestion
Interoperating between the two module systems brings large challenges, with many new features to juggle;
however, support for ESM in Node.js is now implemented in Node.js, and the dust has begun to settle.
```

This setting controls whether `.js` files are interpreted as ES modules or CommonJS modules, and defaults to CommonJS when not set.
When a file is considered an ES module, a few different rules come into play compared to CommonJS:

> * `import`/`export` statements, and top-level `await` can be used

Reviewer 2: ```suggestion
* `import`/`export` statements and top-level `await` can be used
```

* relative import paths need full extensions (e.g we have to write `import "./foo.js"` instead of `import "./foo"`)
* imports might resolve differently from dependencies in `node_modules`

> * certain global-like values like `require()`, `process` and `__dirname` cannot be used directly

Reviewer 2: `process` was always wrong here

<img width="412" alt="image" src="https://user-images.githubusercontent.com/3277153/175375364-a8b85314-99bd-4d28-8ab0-01ffe3284a49.png">

```suggestion
* certain global-like values like `require()` and `__dirname` cannot be used directly
```

This might feel a bit cumbersome at first, but TypeScript tooling like auto-imports and path completion will typically just do this for you.

One other thing to mention is the fact that this applies to `.d.ts` files too.

> When TypeScript finds a `.d.ts` file in package, whether it is treated as an ESM or CommonJS file is interpreted based on the containing package.

Reviewer 2: I feel like this paragraph really needs to be reworked, but since it was in the experimental docs I’ll do a more minimal fix to try to get this out the door

```suggestion
When TypeScript finds a `.d.ts` file in package, whether it is treated as an ESM or CommonJS file is based on the containing package.
```

### New File Extensions

> Using the `type` field in `package.json` is nice way to structure your project because it allows us to continue using the `.ts` and `.js` file extensions which can be convenient;

Reviewer 2: ```suggestion
The `type` field in `package.json` is nice because it allows us to continue using the `.ts` and `.js` file extensions which can be convenient;
```

In some cases, Node.js also synthesizes named exports from CommonJS modules, which can be more convenient.
In these cases, ES modules can use a "namespace-style" import (i.e. `import * as foo from "..."`), or named imports (i.e. `import { helper } from "..."`).

```ts twoslash
// @module: nodenext
// @filename: helper.cts
export function helper() {
    console.log("hello world!");
}

// @filename: index.mts

> import { helper } from "./v.cjs";

Reviewer 2: ```suggestion
import { helper } from "./helper.cjs";
```

?

// prints "hello world!"
helper();
```

There isn't always a way for TypeScript to know whether these named imports will be synthesized, but TypeScript will err on being permissive and use some heuristics when importing from a file that is definitely a CommonJS module.

> One _TypeScript-specific_ note about interop is the following syntax:

Reviewer 2: ```suggestion
One TypeScript-specific note about interop is the following syntax:
```
